File: PullRefresh/src/main/java/com/yan/pullrefreshlayout/BezierInterpolator.java
package com.yan.pullrefreshlayout;

/**
 * @author kaibo
 * @date 2018/10/19 14:34
 * @GitHub：https://github.com/yuxuelian
 * @email：kaibo1hao@gmail.com
 * @description：
 */

import android.graphics.PointF;
import android.view.animation.Interpolator;

/**
 * @author 56896
 * @date 2018/10/18 23:54
 * @GitHub：https://github.com/yuxuelian
 * @email：kaibo1hao@gmail.com
 * @description：
 */
public class BezierInterpolator implements Interpolator {

    private static final float STEP_SIZE = 1.0f / 4096;
    private final PointF point1 = new PointF();
    private final PointF point2 = new PointF();

    private int mLastI = 0;

    public BezierInterpolator(float x1, float y1, float x2, float y2) {
        point1.x = x1;
        point1.y = y1;
        point2.x = x2;
        point2.y = y2;
    }

    @Override
    public float getInterpolation(float input) {
        float t = input;
        //如果重新开始要重置缓存的i。
        if ((int) (input * 10) == 0) {
            mLastI = 0;
        }
        // 近似求解t
        double tempX;
        for (int i = mLastI; i < 4096; i++) {
            t = i * STEP_SIZE;
            tempX = cubicEquation(t, point1.x, point2.x);
            if (tempX >= input) {
                mLastI = i;
                break;
            }
        }
        double value = cubicEquation(t, point1.y, point2.y);
        return (float) value;
    }

    public static double cubicEquation(double t, double p1, double p2) {
        double u = 1 - t;
        double tt = t * t;
        double uu = u * u;
        double ttt = tt * t;
        return 3 * uu * t * p1 + 3 * u * tt * p2 + ttt;
    }
}

File: PullRefresh/src/main/java/com/yan/pullrefreshlayout/BezierInterpolator.kt
package com.yishi.refresh

import android.graphics.PointF
import android.view.animation.Interpolator

/**
 * @author 56896
 * @date 2018/10/18 23:54
 * @GitHub：https://github.com/yuxuelian
 * @email：kaibo1hao@gmail.com
 * @description：
 */
class BezierInterpolator(x1: Float, y1: Float, x2: Float, y2: Float) : Interpolator {
    private val point1 = PointF()
    private val point2 = PointF()

    private var mLastI = 0

    init {
        point1.x = x1
        point1.y = y1
        point2.x = x2
        point2.y = y2
    }

    constructor() : this(.25f, .46f, .45f, .94f)

    override fun getInterpolation(input: Float): Float {
        var t = input
        //如果重新开始要重置缓存的i。
        if ((input * 10).toInt() == 0) {
            mLastI = 0
        }
        // 近似求解t
        var tempX: Double
        for (i in mLastI..4095) {
            t = i * STEP_SIZE
            tempX = cubicEquation(t.toDouble(), point1.x.toDouble(), point2.x.toDouble())
            if (tempX >= input) {
                mLastI = i
                break
            }
        }
        val value = cubicEquation(t.toDouble(), point1.y.toDouble(), point2.y.toDouble())
        return value.toFloat()
    }

    companion object {

        private const val STEP_SIZE = 1.0f / 4096

        fun cubicEquation(t: Double, p1: Double, p2: Double): Double {
            val u = 1 - t
            val tt = t * t
            val uu = u * u
            val ttt = tt * t
            return 3.0 * uu * t * p1 + 3.0 * u * tt * p2 + ttt
        }
    }
}

